src/navigator.rs

```diff
@@ -0,0 +1,154 @@
+    views: Views<H, T>,
+}
+type Views<H, T> = HashMap<H, Box<dyn Fn() -> Box<dyn Widget<T>>>>;
+
+impl<T: Data, H: View> Navigator<T, H> {
+    /// Creates new Navigator widget with the initial view.
+    pub fn new(name: H, ui_builder: impl Fn() -> Box<dyn Widget<T>> + 'static) -> Self {
+        let mut views = Vec::new();
+        let current_view = (ui_builder)();
+        let current_view = WidgetPod::new(current_view);
+        views.push(current_view);
+        let mut this = Self {
+            state: views,
+            views: HashMap::new(),
+        };
+        if this.views.insert(name, Box::new(ui_builder)).is_some() {
+            unreachable!("Map should be empty at this point");
+        }
+        this
+    }
+
+    /// Takes a function to build a widget and a name that will be used to refer to it.
+    pub fn with_view_builder(
+        mut self,
+        name: H,
+        view_builder: impl Fn() -> Box<dyn Widget<T>> + 'static,
+    ) -> Self {
+        if self.views.insert(name, Box::new(view_builder)).is_some() {
+            log::warn!("Views should never update. They should be set at navigator creation and never change.");
+        }
+        self
+    }
+
+    /// Pushes a new view into navigator's state to be displayed
+    fn push_view(&mut self, view: H) {
+        let ui_builder = self.views.get(&view).unwrap();
+        let new_view = (ui_builder)();
+        let widget = WidgetPod::new(new_view);
+        self.state.push(widget);
+    }
+
+    /// Removes a view from navigator's state
+    fn pop_view(&mut self) {
+        if self.state.len() == 1 {
+            return;
+        }
+        self.state.pop().unwrap();
+    }
+}
+/// This gives your Application State the behavior necessary to manipulate its views.
+///
+/// You will want to have your AppState or any State implement this so the navigator
+/// can get relevant state to update its child widgets.
+pub trait ViewController<T: Hash + PartialEq + Eq + Clone> {
+    /// Pushes a new view to be displayed.
+    fn add_view(&mut self, view: T);
+    /// Removes the current view from display.
+    fn pop_view(&mut self);
+    /// Gets the current view that is being displayed.
+    fn current_view(&self) -> &T;
+    /// Gets the length of the backing View data structure.
+    ///
+    /// Views will probably be backed by some kind of array.
+    fn len(&self) -> usize;
+    /// If views is empty then this should be an error.
+    fn is_empty(&self) -> bool;
+}
+
+/// A view will act as representation for the child widget within Navigator.
+pub trait View: Hash + PartialEq + Eq + Clone + fmt::Debug {}
+
+impl<H: View, T: Data + ViewController<H>> Widget<T> for Navigator<T, H> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        // each child view gets to handle the event before a view might be popped
+        if event.should_propagate_to_hidden() {
+            for view in self.state.iter_mut() {
```

Comment:
This needs to check if it should propagate events using Event::should_propagate_to_hidden